java/rocksjni/jni_merge_operator_v2.cc

```diff
@@ -0,0 +1,232 @@
+//
+// Created by rhubner on 29-Nov-23.
+//
+
+#include "jni_merge_operator_v2.h"
+
+#include "include/org_rocksdb_MergeOperatorV2.h"
+#include "rocksjni/cplusplus_to_java_convert.h"
+#include "rocksjni/portal.h"
+
+jlong Java_org_rocksdb_MergeOperatorV2_toCString(JNIEnv* env, jclass,
+                                                 jstring operator_name) {
+  auto operator_name_utf = env->GetStringUTFChars(operator_name, nullptr);
+  if (operator_name_utf == nullptr) {
+    return 0;  // Exception
+  }
+  auto operator_name_len = env->GetStringUTFLength(operator_name);
+
+  char* ret_value = new char[operator_name_len + 1];
+  memcpy(ret_value, operator_name_utf, operator_name_len + 1);
+
+  env->ReleaseStringUTFChars(operator_name, operator_name_utf);
+
+  return GET_CPLUSPLUS_POINTER(ret_value);
+}
+
+jlong Java_org_rocksdb_MergeOperatorV2_newMergeOperator(
+    JNIEnv* env, jobject java_merge_operator, jlong _operator_name) {
+  char* operator_name = reinterpret_cast<char*>(_operator_name);
+
+  auto* jni_merge_operator =
+      new std::shared_ptr<ROCKSDB_NAMESPACE::MergeOperator>(
+          new rocksdb::JniMergeOperatorV2(env, java_merge_operator,
+                                          operator_name));
+
+  return GET_CPLUSPLUS_POINTER(jni_merge_operator);
+}
+
+void Java_org_rocksdb_MergeOperatorV2_disposeInternal(JNIEnv*, jclass,
+                                                      jlong j_handle) {
+  auto* jni_merge_operator =
+      reinterpret_cast<std::shared_ptr<ROCKSDB_NAMESPACE::MergeOperator>*>(
+          j_handle);
+  delete jni_merge_operator;
+}
+
+namespace ROCKSDB_NAMESPACE {
+
+JniMergeOperatorV2::JniMergeOperatorV2(JNIEnv* env, jobject java_merge_operator,
+                                       char* _operator_name)
+    : JniCallback(env, java_merge_operator) {
+  operator_name = _operator_name;
+
+  j_merge_class = env->GetObjectClass(java_merge_operator);
+  if (j_merge_class == nullptr) {
+    return;  // Exception
+  }
+  j_merge_class = static_cast<jclass>(env->NewGlobalRef(j_merge_class));

```

Comment:
The JNI manual says NewGlobalRef() can return nullptr WITHOUT an exception, e.g. if the system is out of memory. So we should throw a Java exception here. Can you also check the other JNI methods you call, to confirm that they always set an exception on nullptr, and if they don't we should set one in each case.

Then fix the comments to say "XYZException has been thrown", or whatever is appropriate.

Reply:
I check existing code and we don't throw exceptions when we can't create global references.

Reply:
Really ? That seems wrong, and may be an omission. If NewGlobalRef() is returning null, surely the right thing to do is to throw an exception, unless we somehow cope with j_merge_class as null ? So let us do it correctly here for now.

Reply:
@alanpaxton Fixed.

Reply:
Thanks. I'm happy.